File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/engine/tier/TierConfigTest.kt
package org.wyrdsekai.app.engine.tier

import kotlin.test.Test
import kotlin.test.assertTrue
import kotlin.test.assertEquals

class TierConfigTest {

    @Test
    fun tierConfigsAreConsistentWithSpec() {
        // T0: soul + nexus (companion needs a room), remote inference
        val t0 = TierConfig.forTier(Tier.T0)
        assertEquals(0, t0.maxRooms) // maxRooms is for non-essential rooms; nexus is always-on
        assertEquals(InferenceMode.REMOTE, t0.inferenceMode)
        assertEquals(false, t0.betweenEnabled)

        // T1: one room, local inference
        val t1 = TierConfig.forTier(Tier.T1)
        assertEquals(1, t1.maxRooms)
        assertEquals(InferenceMode.LOCAL, t1.inferenceMode)
        assertEquals(false, t1.betweenEnabled)

        // T2: multi-room, Between enabled
        val t2 = TierConfig.forTier(Tier.T2)
        assertEquals(4, t2.maxRooms)
        assertEquals(true, t2.betweenEnabled)

        // T3: full peer
        val t3 = TierConfig.forTier(Tier.T3)
        assertEquals(16, t3.maxRooms)
        assertEquals(2, t3.maxConcurrentInference)
        assertEquals(true, t3.betweenEnabled)
    }

    @Test
    fun contextTokensScaleWithTier() {
        val tiers = Tier.entries.map { TierConfig.forTier(it).maxContextTokens }
        // Each tier should have >= previous tier's tokens
        for (i in 1 until tiers.size) {
            assertTrue(tiers[i] >= tiers[i - 1], "T${i} context tokens should be >= T${i - 1}")
        }
    }

    // ── ResourceSnapshot tier recommendation ────────────────────────────

    @Test
    fun criticalThermalForcesT0() {
        val snap = ResourceSnapshot(
            availableMemoryMb = 4000,
            totalMemoryMb = 8000,
            batteryPercent = 100,
            isCharging = true,
            thermalState = ThermalState.CRITICAL,
            hasWifi = true,
        )
        assertEquals(Tier.T0, snap.recommendTier())
    }

    @Test
    fun lowBatteryNotChargingForcesT0() {
        val snap = ResourceSnapshot(
            availableMemoryMb = 4000,
            totalMemoryMb = 8000,
            batteryPercent = 5,
            isCharging = false,
            thermalState = ThermalState.NOMINAL,
            hasWifi = true,
        )
        assertEquals(Tier.T0, snap.recommendTier())
    }

    @Test
    fun chargingWithWifiAndHighMemoryGivesT3() {
        val snap = ResourceSnapshot(
            availableMemoryMb = 4000,
            totalMemoryMb = 8000,
            batteryPercent = 80,
            isCharging = true,
            thermalState = ThermalState.NOMINAL,
            hasWifi = true,
        )
        assertEquals(Tier.T3, snap.recommendTier())
    }

    @Test
    fun wifiWithGoodResourcesGivesT2() {
        val snap = ResourceSnapshot(
            availableMemoryMb = 2500,
            totalMemoryMb = 4000,
            batteryPercent = 60,
            isCharging = false,
            thermalState = ThermalState.NOMINAL,
            hasWifi = true,
        )
        assertEquals(Tier.T2, snap.recommendTier())
    }

    @Test
    fun moderateResourcesGivesT1() {
        val snap = ResourceSnapshot(
            availableMemoryMb = 1500,
            totalMemoryMb = 3000,
            batteryPercent = 40,
            isCharging = false,
            thermalState = ThermalState.NOMINAL,
            hasWifi = false,
        )
        assertEquals(Tier.T1, snap.recommendTier())
    }

    @Test
    fun lowMemoryGivesT0() {
        val snap = ResourceSnapshot(
            availableMemoryMb = 800,
            totalMemoryMb = 2000,
            batteryPercent = 40,
            isCharging = false,
            thermalState = ThermalState.NOMINAL,
            hasWifi = true,
        )
        assertEquals(Tier.T0, snap.recommendTier())
    }

    @Test
    fun lowBatteryWhileChargingStillAllowsT1() {
        val snap = ResourceSnapshot(
            availableMemoryMb = 2000,
            totalMemoryMb = 4000,
            batteryPercent = 5,
            isCharging = true,
            thermalState = ThermalState.NOMINAL,
            hasWifi = false,
        )
        assertEquals(Tier.T1, snap.recommendTier())
    }
}

File: clients/rn/__tests__/engine/tier-config.test.ts
import {
  type ResourceSnapshot,
  TIER_ORDER,
  configForTier,
  recommendTier,
  tierIndex,
} from '../../src/engine/tier/TierConfig';

describe('TierConfig', () => {
  it('tier configs are consistent with spec', () => {
    // T0: remote inference, no non-essential rooms
    const t0 = configForTier('T0');
    expect(t0.maxRooms).toBe(0);
    expect(t0.inferenceMode).toBe('REMOTE');
    expect(t0.betweenEnabled).toBe(false);

    // T1: one room, local inference
    const t1 = configForTier('T1');
    expect(t1.maxRooms).toBe(1);
    expect(t1.inferenceMode).toBe('LOCAL');
    expect(t1.betweenEnabled).toBe(false);

    // T2: multi-room, Between enabled
    const t2 = configForTier('T2');
    expect(t2.maxRooms).toBe(4);
    expect(t2.betweenEnabled).toBe(true);

    // T3: full peer
    const t3 = configForTier('T3');
    expect(t3.maxRooms).toBe(16);
    expect(t3.maxConcurrentInference).toBe(2);
    expect(t3.betweenEnabled).toBe(true);
  });

  it('context tokens scale with tier', () => {
    const tokens = TIER_ORDER.map(t => configForTier(t).maxContextTokens);
    for (let i = 1; i < tokens.length; i++) {
      expect(tokens[i]).toBeGreaterThanOrEqual(tokens[i - 1]);
    }
  });

  it('tier index is ordered', () => {
    expect(tierIndex('T0')).toBe(0);
    expect(tierIndex('T1')).toBe(1);
    expect(tierIndex('T2')).toBe(2);
    expect(tierIndex('T3')).toBe(3);
  });
});

describe('ResourceSnapshot recommendTier', () => {
  const snap = (overrides: Partial<ResourceSnapshot> = {}): ResourceSnapshot => ({
    availableMemoryMb: 2000,
    totalMemoryMb: 4000,
    batteryPercent: 80,
    isCharging: false,
    thermalState: 'NOMINAL',
    hasWifi: true,
    ...overrides,
  });

  it('critical thermal forces T0', () => {
    expect(recommendTier(snap({
      availableMemoryMb: 4000,
      batteryPercent: 100,
      isCharging: true,
      thermalState: 'CRITICAL',
    }))).toBe('T0');
  });

  it('low battery not charging forces T0', () => {
    expect(recommendTier(snap({
      availableMemoryMb: 4000,
      batteryPercent: 5,
      isCharging: false,
    }))).toBe('T0');
  });

  it('charging with wifi and high memory gives T3', () => {
    expect(recommendTier(snap({
      availableMemoryMb: 4000,
      batteryPercent: 80,
      isCharging: true,
      hasWifi: true,
    }))).toBe('T3');
  });

  it('wifi with good resources gives T2', () => {
    expect(recommendTier(snap({
      availableMemoryMb: 2500,
      batteryPercent: 60,
      isCharging: false,
      hasWifi: true,
    }))).toBe('T2');
  });

  it('moderate resources gives T1', () => {
    expect(recommendTier(snap({
      availableMemoryMb: 1500,
      batteryPercent: 40,
      isCharging: false,
      hasWifi: false,
    }))).toBe('T1');
  });

  it('low memory gives T0', () => {
    expect(recommendTier(snap({
      availableMemoryMb: 800,
      batteryPercent: 40,
      isCharging: false,
    }))).toBe('T0');
  });

  it('low battery while charging still allows T1', () => {
    expect(recommendTier(snap({
      availableMemoryMb: 2000,
      batteryPercent: 5,
      isCharging: true,
      hasWifi: false,
    }))).toBe('T1');
  });
});
